### DISCUSSION TITLE
Output of drawings

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1685] [Looking for help]
I have extracted drawings using page.get_drawings(). Can you let me know how to extract the values that are present in the content stream without any transformation?

![Capture](https://user-images.githubusercontent.com/36301306/164517789-4b05b4bc-3afe-4381-ba01-31cc14ae74e7.JPG)
For example in the content stream the rect values are 100 700 100 100 whereas the rect value obtained using get_drawings had values 
100 42 200 142.

Can you let me know how to obtain the same rect values present in the content stream?

**Comment by JorjMcKie** [comment]
In (Py-) MuPDF, coordinates count from top to bottom: top-left has coordinates (0, 0).
In PDF, page **_bottom_**-left has (0, 0).
So you need the `ptm = page.transformation_matrix` to translate between these coordinate systems. Given a PDF point (x, y), then `Point(x, y) * ptm` gives the point in MuPDF coordinates. The other way round, a point in MuPDF coordinates (as contained in `.get_drawings()`) has ccordinates `Point(x, y) * ~ptm` in PDF.

**Reply by JorjMcKie** [reply]
But keep in mind, that values seen in the content stream may be subject to multiple matrix transformations (`"cm"`) coded before, like this
```postscript
1 2 3 4 5 6 cm
...
6 5 4 3 2 1 cm
...
100 700 100 100 re
...
```
So MuPDF sequentially walks through all those `cm` instructions, multiplies these matrices one by one and finally with the ptm and then the rectangle coordinates to give the MuPDF values.

**Reply by meghanaviyyapu** [reply]
Is there difference in drawings creation when page.getcdrawings() is used instead of page.get_drawings()?

**Reply by JorjMcKie** [reply]
Except for a great performance difference, the C-version only delivers tuples, no fitz.Point, fitz.Rect, fitz.Quad.

**Comment by meghanaviyyapu** [comment]
Can you let me know if the blue drawing behind some text in the PDF can be extracted using PyMuPDF?
[sample_pdf.pdf](https://github.com/pymupdf/PyMuPDF/files/12157504/sample_pdf.pdf)

**Reply by JorjMcKie** [reply]
These are vector graphics, so you can extract them using `page.get_drawings()`.
However, the fill color is shaded: this is **_not_** currently supported.

**Reply by meghanaviyyapu** [reply]
Thanks. Is there any way to get the shaded part using PyMuPDF?

**Comment by JorjMcKie** [comment]
> Thanks. Is there any way to get the shaded part using PyMuPDF?

Sorry, not currently.
